File: android/app/src/main/kotlin/com/chipbase/chipcore_sdk/flutter/BlockchainSpec.kt
package com.chipcore.sdk.flutter

import android.util.Log
import java.nio.charset.StandardCharsets
import java.util.Locale

internal sealed class BlockchainSpec(
    val coinType: Int,
    /** 对应 card_coin Blockchain.id，用于 WalletManagerRegistry 的键 */
    val id: String,
) {
    abstract fun matches(identifier: String): Boolean
    abstract fun makeAddress(publicKey: ByteArray, isTest: Boolean): String

    open fun resolveDigest(payload: String): ByteArray {
        return payload.hexToByteArrayOrNull()?.takeIf { it.size == 32 }
            ?: sha256(payload.toByteArray(StandardCharsets.UTF_8))
    }

    /**
     * 生成 BIP44/BIP84 路径字节。子类可重写以使用不同 purpose。
     * change=-1 表示路径截止到 account 段（3段：m/purpose'/coinType'/account'）。
     * index=-1  表示路径截止到 change 段（4段：m/purpose'/coinType'/account'/change）。
     */
    open fun defaultPath(account: Int = 0, change: Int = 0, index: Int = 0): ByteArray {
        val segments = when {
            change < 0 -> intArrayOf(44 or HARDENED, coinType or HARDENED, account or HARDENED)
            index < 0  -> intArrayOf(44 or HARDENED, coinType or HARDENED, account or HARDENED, change)
            else       -> intArrayOf(44 or HARDENED, coinType or HARDENED, account or HARDENED, change, index)
        }
        val output = ByteArray(segments.size * 4)
        segments.forEachIndexed { idx, value ->
            val offset = idx * 4
            output[offset] = ((value ushr 24) and 0xFF).toByte()
            output[offset + 1] = ((value ushr 16) and 0xFF).toByte()
            output[offset + 2] = ((value ushr 8) and 0xFF).toByte()
            output[offset + 3] = (value and 0xFF).toByte()
        }
        return output
    }

    object Bitcoin : BlockchainSpec(0, "btc") {
        override fun matches(identifier: String): Boolean {
            val value = identifier.lowercase(Locale.US)
            return value.contains("btc") || value.contains("bitcoin")
        }

        // 路径使用父类默认 m/44'/0'/0'/0/0，地址格式为 P2WPKH bech32 (bc1q/tb1q)
        override fun makeAddress(publicKey: ByteArray, isTest: Boolean): String {
            val compressed = publicKey.compressSecp256k1()
            val hash160 = ripemd160(sha256(compressed))
            val hrp = if (isTest) "tb" else "bc"
            return Bech32.encodeP2WPKH(hrp, hash160)
        }
    }

    object Ethereum : BlockchainSpec(1, "eth") {
        override fun matches(identifier: String): Boolean {
            val value = identifier.lowercase(Locale.US)
            return value.contains("eth") || value.contains("ethereum") || value.contains("evm")
        }

        override fun makeAddress(publicKey: ByteArray, isTest: Boolean): String {
            // 卡片返回的公钥可能是压缩格式(33字节)或未压缩格式(65字节)或纯 X+Y(64字节)
            // ETH 地址 = keccak256(未压缩 X+Y 64字节) 的后 20 字节
            val ethBytes = publicKey.decompressToEthBytes()
            Log.d("ChipCoreNfc", "ETH makeAddress: pubKey[${publicKey.size}]=${publicKey.toHex()} -> ethBytes[${ethBytes.size}]")
            val hash = keccak256(ethBytes)
            val rawHex = hash.copyOfRange(hash.size - 20, hash.size).toHex()
            return "0x" + eip55ChecksumAddress(rawHex)
        }

        override fun resolveDigest(payload: String): ByteArray {
            return payload.hexToByteArrayOrNull()?.takeIf { it.size == 32 }
                ?: keccak256(payload.toByteArray(StandardCharsets.UTF_8))
        }
    }

    /** TRON (TRX) — BIP44 coinType=195, path m/44'/195'/0'/0/0 */
    object Tron : BlockchainSpec(195, "trx") {
        override fun matches(identifier: String): Boolean {
            val value = identifier.lowercase(Locale.US)
            return value.contains("trx") || value.contains("tron")
        }

        override fun makeAddress(publicKey: ByteArray, isTest: Boolean): String {
            // TRX 地址 = Base58Check(0x41 + keccak256(uncompressedPubKey)[12:32])
            val ethBytes = publicKey.decompressToEthBytes()
            val hash = keccak256(ethBytes)
            val raw20 = hash.copyOfRange(hash.size - 20, hash.size)
            val payload = byteArrayOf(0x41.toByte()) + raw20
            Log.d("ChipCoreNfc", "TRX makeAddress: payload=${payload.toHex()}")
            return base58Check(payload)
        }

        override fun resolveDigest(payload: String): ByteArray {
            return payload.hexToByteArrayOrNull()?.takeIf { it.size == 32 }
                ?: keccak256(payload.toByteArray(StandardCharsets.UTF_8))
        }
    }

    /** Dogecoin (DOGE) — BIP44 coinType=3, path m/44'/3'/0'/0/0 */
    object Dogecoin : BlockchainSpec(3, "doge") {
        override fun matches(identifier: String): Boolean {
            val value = identifier.lowercase(Locale.US)
            return value.contains("doge") || value.contains("dogecoin")
        }

        override fun makeAddress(publicKey: ByteArray, isTest: Boolean): String {
            // DOGE P2PKH: version=0x1E(mainnet)/0x71(testnet), hash160(compressedPubKey)
            val compressed = publicKey.compressSecp256k1()
            val hash160 = ripemd160(sha256(compressed))
            val version = if (isTest) 0x71.toByte() else 0x1E.toByte()
            val payload = byteArrayOf(version) + hash160
            Log.d("ChipCoreNfc", "DOGE makeAddress: isTest=$isTest payload=${payload.toHex()}")
            return base58Check(payload)
        }
    }

    companion object {
        private const val HARDENED = 0x80000000.toInt()

        fun fromCurrency(currency: Messages.CurrencyInfoMessage): BlockchainSpec {
            return fromIdentifier(listOf(currency.networkId, currency.symbol, currency.name).joinToString(" "))
        }

        fun fromIdentifier(identifier: String): BlockchainSpec {
            return when {
                Bitcoin.matches(identifier)   -> Bitcoin
                Tron.matches(identifier)      -> Tron
                Dogecoin.matches(identifier)  -> Dogecoin
                else                          -> Ethereum
            }
        }
    }
}

File: ios/Runner/BlockchainSpec.swift
import CryptoSwift
import Flutter
import Foundation

// MARK: - BlockchainSpec（对应 Android BlockchainSpec.kt）

enum BlockchainSpec {
  case bitcoin
  case ethereum
  case tron
  case dogecoin

  /// 对应 Android BlockchainSpec.id，用于 WalletManagerRegistry 的键
  var id: String {
    switch self {
    case .bitcoin:  return "btc"
    case .ethereum: return "eth"
    case .tron:     return "trx"
    case .dogecoin: return "doge"
    }
  }

  func matches(_ identifier: String) -> Bool {
    let value = identifier.lowercased()
    switch self {
    case .bitcoin:
      return value.contains("btc") || value.contains("bitcoin")
    case .ethereum:
      return value.contains("eth") || value.contains("ethereum") || value.contains("evm")
    case .tron:
      return value.contains("trx") || value.contains("tron")
    case .dogecoin:
      return value.contains("doge") || value.contains("dogecoin")
    }
  }

  static func fromCurrency(_ currency: CurrencyInfoMessage) -> BlockchainSpec {
    fromIdentifier([currency.networkId, currency.symbol, currency.name].joined(separator: " "))
  }

  static func fromIdentifier(_ identifier: String) -> BlockchainSpec {
    if bitcoin.matches(identifier) { return .bitcoin }
    if tron.matches(identifier) { return .tron }
    if dogecoin.matches(identifier) { return .dogecoin }
    return .ethereum
  }

  func defaultPath(account: UInt32 = 0, change: UInt32 = 0, index: UInt32 = 0) -> Data {
    let hardened: UInt32 = 0x80000000
    let coinType: UInt32
    switch self {
    case .bitcoin:  coinType = 0
    case .tron:     coinType = 195
    case .dogecoin: coinType = 3
    case .ethereum: coinType = 1   // 与 Android BlockchainSpec.Ethereum(coinType=1) 保持一致，路径 m/44'/1'/0'/0/0
    }
    let segments: [UInt32] = [44 | hardened, coinType | hardened, account | hardened, change, index]
    return segments.reduce(into: Data()) { result, segment in
      var bigEndian = segment.bigEndian
      result.append(Data(bytes: &bigEndian, count: MemoryLayout<UInt32>.size))
    }
  }

  func resolveDigest(_ payload: String) -> Data {
    if let data = payload.hexToData(), data.count == 32 {
      return data
    }
    let bytes = Array(payload.utf8)
    switch self {
    case .bitcoin, .dogecoin:
      return Data(bytes.sha256())
    case .ethereum, .tron:
      return Data(bytes.sha3(.keccak256))
    }
  }

  func makeAddress(publicKey: Data, isTest: Bool) -> String {
    switch self {
    case .bitcoin:
      let compressed = publicKey.compressedSecp256k1()
      let hash160 = Data(compressed.bytes.sha256().ripemd160())
      let hrp = isTest ? "tb" : "bc"
      return Bech32.encodeP2WPKH(hrp: hrp, hash160: hash160)
    case .dogecoin:
      let compressed = publicKey.compressedSecp256k1()
      let hash160 = Data(compressed.bytes.sha256().ripemd160())
      let prefix = Data([isTest ? 0x71 : 0x1E])
      let body = prefix + hash160
      let checksum = Data(Data(body.bytes.sha256()).bytes.sha256().prefix(4))
      return Base58.encode(body + checksum)
    case .tron:
      let normalized = publicKey.count == 65 && publicKey.first == 0x04 ? publicKey.dropFirst() : Data(publicKey)
      let hash = Data(Array(normalized).sha3(.keccak256))
      let body = Data([0x41]) + Data(hash.suffix(20))
      let checksum = Data(Data(body.bytes.sha256()).bytes.sha256().prefix(4))
      return Base58.encode(body + checksum)
    case .ethereum:
      let normalized = publicKey.count == 65 && publicKey.first == 0x04 ? publicKey.dropFirst() : Data(publicKey)
      let hash = Data(Array(normalized).sha3(.keccak256))
      let rawHex = Data(hash.suffix(20)).hexString
      return "0x" + eip55Checksum(rawHex)
    }
  }
}

// MARK: - Base58

enum Base58 {
  private static let alphabet = Array("123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz")

  static func encode(_ data: Data) -> String {
    if data.isEmpty { return "" }
    var bytes = Array(data)
    let zeroCount = bytes.prefix { $0 == 0 }.count
    var startAt = zeroCount
    var encoded: [Character] = []

    while startAt < bytes.count {
      var remainder = 0
      for index in startAt..<bytes.count {
        let value = Int(bytes[index]) & 0xFF
        let temp = remainder * 256 + value
        bytes[index] = UInt8(temp / 58)
        remainder = temp % 58
      }
      encoded.append(alphabet[remainder])
      while startAt < bytes.count && bytes[startAt] == 0 {
        startAt += 1
      }
    }

    for _ in 0..<zeroCount {
      encoded.append(alphabet[0])
    }
    return String(encoded.reversed())
  }
}

// MARK: - CurrencyInfoMessage extension

extension CurrencyInfoMessage {
  func copyWith(publicKey: Data? = nil, chainCode: Data? = nil, address: String? = nil) -> CurrencyInfoMessage {
    CurrencyInfoMessage(
      id: id,
      icon: icon,
      name: name,
      networkId: networkId,
      networkName: networkName,
      networkIcon: networkIcon,
      symbol: symbol,
      contractAddress: contractAddress,
      decimalCount: decimalCount,
      amount: amount,
      address: address ?? self.address,
      publicKey: publicKey.map { FlutterStandardTypedData(bytes: $0) } ?? self.publicKey,
      chainCode: chainCode.map { FlutterStandardTypedData(bytes: $0) } ?? self.chainCode,
      isTest: isTest
    )
  }
}

// MARK: - EIP-55 混合大小写校验和

private func eip55Checksum(_ hexNoPrefix: String) -> String {
  let lower = hexNoPrefix.lowercased()
  let hashBytes = Array(lower.utf8).sha3(.keccak256)
  var result = ""
  for (i, c) in lower.enumerated() {
    let nibble = (Int(hashBytes[i / 2]) >> (i % 2 == 0 ? 4 : 0)) & 0xF
    result.append(nibble >= 8 && c.isLetter ? Character(c.uppercased()) : c)
  }
  return result
}
